app/services/work_packages/set_schedule_service.rb

```diff
@@ -174,4 +178,34 @@ def set_dates(work_package, start_date, due_date)
   def days(work_package)
     WorkPackages::Shared::Days.for(work_package)
   end
+
+  def assign_cause_for_journaling(work_package, relation)
+    return {} if initiated_by.nil?
+
+    if initiated_by.is_a?(WorkPackage)
+      assign_cause_initiated_by_work_package(work_package, relation)
+    elsif initiated_by.is_a?(Journal::WorkingDayUpdate)
+      assign_cause_initiated_by_changed_working_days(work_package)
+    end
+  end
+
+  def assign_cause_initiated_by_work_package(work_package, relation)
+    type_mapping = {
+      parent: 'work_package_parent_changed_times',
+      successor: 'work_package_successor_changed_times',
+      predecessor: 'work_package_predecessor_changed_times'
+    }
+
+    work_package.journal_cause = {
+      "type" => type_mapping[relation],
+      "work_package_id" => initiated_by.id
+    }
+  end
+
+  def assign_cause_initiated_by_changed_working_days(work_package)
+    work_package.journal_cause = {
+      "type" => 'working_days_changed',
+      "changed_days" => initiated_by.to_h
+    }
+  end
 end
```

Comment:
Given the sometimes hidden relationship (e.g. caused by a change to the predecessor or to a predecessor of an ancestor, it might be simpler to stick to finding out the kind of relationship not through the code paths but rather by the reference itself.

Since there will always only be one work package causing it, one could find out the direct predecessors, children and the parent for this work package (or at least their ids) and when setting the cause, check if the altered work package is in any of the sets and if that is the case, use the key for the set it is found in.. If it isn't, the general fallback will be used.

Reply:
You mean that **date changed by changes to predecessor 123** is only used in the direct followers and not in the entire chain?